More failure logs

xu-cheng opened the pull request:
cc @mikemcquaid

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1288,9 +1290,18 @@ def system(cmd, *args)
+              log.puts buf
+              puts buf
+            end
+          else # travis
```
I think we should do this always when `--verbose` isn't passed, actually.

mikemcquaid replied:
e.g. don't limit it just to travis. I think it would help avoid confusion for long builds like `boost` (as long as we keep the dot count sensible).

xu-cheng replied:
In that case, we need to introduce a new flag, `--semi-verbose` or something.

mikemcquaid replied:
I don't agree. Users are confused with the current behaviour and I think this would be an improvement. If you want you can make it `HOMEBREW_DEVELOPER` only for now but I do think it's a sensible change to make.

mikemcquaid replied:
I really don't like the idea of making `system` behave differently for just one CI system.

xu-cheng replied:
> I really don't like the idea of making system behave differently for just one CI system.

That's exactly what I'm talking about. We use a flag here and set this flag based on CI system in `test-bot`.

mikemcquaid commented on Library/Homebrew/cmd/test-bot.rb:
```diff
@@ -155,9 +154,12 @@ def run
+            print buf
+            $stdout.flush
+          end
+          @output << buf
```
What's the reason/advantage of this change?

xu-cheng replied:
The original `read.gets` will only return for every new line. When we print dots, they will be all in one single line, hence won't be printed out.

mikemcquaid replied:
Ah, I see. Would be good to indicate what the `32` is by splitting it out into a variable. Also, it seems like `32` is probably more bytes than we want if we want to print every dot.

xu-cheng replied:
We cannot make the variable too low as it will harm the performance.

mikemcquaid replied:
We could make it lower just for Travis. If we're printing dots based on time output it seems to make sense to ensure that we output every dot otherwise they will e.g. just appear in groups of `16`.

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1288,9 +1290,18 @@ def system(cmd, *args)
+          else # travis
+            while buf = rd.gets
+              log.puts buf
+              print "."
```
It's unclear to me when this will print; after every new write to the output? That seems like it's going to print way too much; it should probably have a timer and print e.g. no more than once a minute.

xu-cheng replied:
Currently, it will print a dot for every line of logs. If you want, we can make it every N lines or every N bytes.

>  it should probably have a timer and print e.g. no more than once a minute.

I kinda like dot printing corresponding real log. So we can see wether the build is really freezing.

mikemcquaid replied:
Yeh, I like the idea of handling build freezing too. I think that printing every line is a good idea but will probably print too many dots. So, instead of printing a dot every 1 minute it prints a dot on every line unless there was already a dot printed in the last minute. How does that sound?

xu-cheng replied:
I don't like the idea of timer too much, it complicates the implementations. Can we just print dot based number of lines?

mikemcquaid replied:
Have you tried this with e.g. `boost`? It seems like it'd produce a vast number of dots.

mikemcquaid replied:
The timer is sensible for Travis because Travis itself is operating with a timer.

xu-cheng replied:
OK, you have convinced me on this.

mikemcquaid replied:
Note I don't mind changing the logic in `test-bot` to be travis-specific or move the timer there.

mikemcquaid replied:
:+1:

mikemcquaid replied:
I'm trying `boost` right now to see what it looks like.

mikemcquaid replied:
```
==> brew install --build-bottle boost
Warning: You are using OS X 10.11.
We do not provide support for this pre-release version.
You may encounter build failures or other breakage.
==> Using the sandbox
==> Downloading https://downloads.sourceforge.net/project/boost/boost/1.58.0/boost_1_58_0.tar.bz2
Already downloaded: /Library/Caches/Homebrew/boost-1.58.0.tar.bz2
==> ./bootstrap.sh --prefix=/usr/local/Cellar/boost/1.58.0 --libdir=/usr/local/Cellar/boost/1.58.0/lib --without-icu --without-libraries=python,mpi
......................
==> ./b2 headers
..........
==> ./b2 --prefix=/usr/local/Cellar/boost/1.58.0 --libdir=/usr/local/Cellar/boost/1.58.0/lib -d2 -j8 --layout=tagged --user-config=user-config.jam install threading=multi,single link=shared,static
...........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................^C
```

:laughing:

xu-cheng commented:
Updated.

mikemcquaid commented on Library/Homebrew/cmd/test-bot.rb:
```diff
@@ -155,9 +154,12 @@ def run
-        while line = read.gets
-          puts line if verbose
-          @output += line
+        while buf = read.read(8)
```
This is 8 bytes (rather than bits) I believe so it's still gonna only print every 8th `.` unfortunately.

mikemcquaid replied:
I really don't think it's worth worrying about performance here given it's a blocking IO operation in a tight loop and only in our CI.

mikemcquaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1288,9 +1290,23 @@ def system(cmd, *args)
+            while buf = rd.gets
+              log.puts buf
+              # make sure dots printed with interval of at least 30 sec.
+              if Time.now - last_dot > 30
```
Does this maybe need to be `(Time.now - last_dot) > 30`? I can never remember the operator precedence.

mikemcquaid replied:
Also, maybe 1 minute rather than 30 seconds?

mikemcquaid commented:
Other than the byte count and operator question: :+1:

xu-cheng commented:
Ok, addressed your feedback.

I also did a small benchmark, using `ruby -e puts("."*10000)` as the test case. It shows `read(1)` will have some slight performance issuing when handling high speed output.
```
Comparison:
                gets:        8.4 i/s
              read 8:        8.4 i/s - 1.00x slower
              read 1:        6.9 i/s - 1.22x slower
```

mikemcquaid commented:
@xu-cheng That seems worth it, to me, given that the actual impact on compilation performance will be minimal and its a relatively small difference.

xu-cheng commented:
I agree with you as such high speed output is kinda rare.

Anything else before :ship: it?

mikemcquaid commented:
:shipit:!
